Refactor frontend for off-chain communication. Many bugs fixed

graup opened the pull request:
- Utility functions (generate state and fen) outsourced into utils file
- Moved chess.js state to game state. Closes #205
- Use chess.js npm package instead of directly including code of library
- Moved event handling from playGame to gamesFactory
- Added last move time to localstorage to redisplay timeout after reload
- Show move number in UI
- Keep track of currently visible game (to enable notifications of moves in other games)
- Fixes #152
- Some other small bugs
- close #210

graup committed:
Save last move time to localstorage to show timeout after reload

graup committed:
oursource fen/state utility functions

graup committed:
Moved chess.js state to games factory and used npm package instead of hardcoded copy of lib

graup committed:
Keep track of visible game to notify user of new moves in other games.

Also fixes #152

graup committed:
show move number in ui

graup committed:
fix sending of move index

kyroy commented:
```
WARNING in ./app/javascript/gamesFactory.js
jshint results in errors
  Line is too long. @ line 475 char 103
        console.log("Sent move no", game.state[8] * 128 + game.state[9], 'from', fromIndex, 'to', toIndex);

  Strings must use singlequote. @ line 475 char 31
        console.log("Sent move no", game.state[8] * 128 + game.state[9], 'from', fromIndex, 'to', toIndex);

  Line is too long. @ line 844 char 101
        if ((game.chess.turn() === 'w' && game.self.color === 'white' && data.args.timeoutState !== 0) ||

  Line is too long. @ line 845 char 101
            (game.chess.turn() === 'b' && game.self.color === 'black' && data.args.timeoutState !== 0)) {

  'generateState' is defined but never used. @ line 3 char 9
    import {generateState, generateFen} from './utils/fen-conversion.js';

WARNING in ./app/javascript/playGame.js
jshint results in errors
  Line is too long. @ line 173 char 112
            gameStates.addSelfMove(game.gameId, algebraicToIndex(move.from), algebraicToIndex(move.to), game.state);
```

graup commented:
There are some new bugs, but I wanted to get this PR up now because I won't have time to look at it further until tomorrow.

Sometimes I get an error that the move counts don't match after reloading. Also the notification for new moves in other games doesn't work 100% yet.

kyroy commented:
Were these bugs also existing before he refactoring?

graup commented:
Before refactoring lots of other things didn't work, so it wasn't really possible to see them. I would say these are new bugs. The basic move, claims / react to claim events and reading from local storage still works.

graup committed:
fix lint errors

kyroy committed:
Use shh from ise.filesmania.de

kyroy commented:
~~`games.update` needs to be changed (see `games.add`)~~  => done

kyroy commented:
```
let lastMoveTime;
try {
  // Try to find out time of last move
  lastMoveTime = gameStates.getLastMoveTime(game);
  if (!lastMoveTime) {
    // get it from blockchain?
  }
} catch (e) {
}
```

The last move time is not stored in the contract. However, we could search the transaction where the last move was made / when the game was joined by the second player

kyroy commented:
When pressing surrender or close, the message "waiting..." is present, but does not get updated. When reloading the page, the action is already completed.

kyroy committed:
Fix bugs / rewrite code for refactoring

- Add games.sendLastStateOrMoveToBlockchain to endgame function calls
- Add some TODOs
- Update games.update function to not overwrite chess objects, but only to update new received data
- Add games.getValidMove function that still has to be implemented
- Remove debug "tab"

kyroy commented:
At some point, we should also remove the move listeners ;)

Commit:
implements getValidMove

kyroy committed:
Revert ELO changes from 2037289

dennis-westphal commented on app/javascript/gamesFactory.js:
```diff
@@ -189,16 +256,27 @@ angular.module('dappChess').factory('games', function (crypto, navigation, gameS
+    for (let i in games.list) {
+      if (games.list[i].gameId === game.gameId) {
+        // update existing game
+        let g = games.list[i];
```
I only used the for loop because I wasn't sure if arr.find would return a reference, but it does; so you can replace the whole loop by let g = games.getGame(game.gameId);

dennis-westphal commented on app/javascript/gamesFactory.js:
```diff
@@ -189,16 +256,27 @@ angular.module('dappChess').factory('games', function (crypto, navigation, gameS
+        let g = games.list[i];
+        for (let key in game) {
+          g[key] = game[key];
+        }
```
use jQuery.extend(true, g, game) instead if you want a deep copy, or jQuery.extend(g, game) if you want to replace sub-objects

dennis-westphal commented on app/javascript/gamesFactory.js:
```diff
@@ -403,6 +484,8 @@ angular.module('dappChess').factory('games', function (crypto, navigation, gameS
       'payload': payload
     });
+    console.log('Sent move no', game.state[8] * 128 + game.state[9],
+                'from', fromIndex, 'to', toIndex);
```
If this is not required anymore, remove it

dennis-westphal commented on app/javascript/gamesFactory.js:
```diff
@@ -431,7 +513,7 @@ angular.module('dappChess').factory('games', function (crypto, navigation, gameS
         }
       }
-    }, 10000);
+    }, 3000);
```
Can you move this to a constant? I'd rather change this for production use later.

dennis-westphal commented on app/javascript/gamesFactory.js:
```diff
@@ -447,23 +529,37 @@ angular.module('dappChess').factory('games', function (crypto, navigation, gameS
+   */
+  games.getValidMove = function (game) {
+    console.log('get valid move for', game);
+    let move = game.chess.moves()[0];
```
I'd rename this to moveFrom or something similar for more clarity

dennis-westphal commented on app/javascript/gamesFactory.js:
```diff
@@ -735,40 +831,84 @@ angular.module('dappChess').factory('games', function (crypto, navigation, gameS
+        (game.chess.turn() === 'w' && game.self.color === 'white') ||
+        (game.chess.turn() === 'b' && game.self.color === 'black'))) {
+      if (
+          ([1, 2].indexOf(data.args.timeoutState) !== -1 && game.chess.in_checkmate()) || // jshint ignore:line
```
Just for understanding: Why does this also apply to a timeout game status? Maybe you could also add this to the comments

kyroy replied:
Added

dennis-westphal commented on app/javascript/gamesFactory.js:
```diff
@@ -735,40 +831,84 @@ angular.module('dappChess').factory('games', function (crypto, navigation, gameS
+          games.sendLastStateOrMoveToBlockchain(game);
+        } catch (e) {
+          $rootScope.$broadcast('message',
+            'Could not send move to blockchain to decline endgame',
```
Please write "end game" instead of "endgame", since it has a different meaning

dennis-westphal commented on static/playGame.html:
```diff
@@ -15,7 +15,7 @@
   <div class="eight columns">
     <p data-ng-show="gameIsActive()">
-      You are in an ongoing chess game, the pot is at {{game.pot}}.
+      You are in an ongoing chess game, the pot is at {{game.pot}}. Move no {{halfMoveNumber}}.
```
Can we please rephrase this? That's not a proper sentence ;-)

kyroy committed:
Fix blames from @dennis-westphal

Use helpers and jQuery.extend in games.update
Remove halving of timeoutTime
Introduce TIMEOUT_ACK constant
Add comments for endgame answers

dennis-westphal commented:
When starting a new game, I get this:

![screen shot 2016-07-12 at 15 38 09](https://cloud.githubusercontent.com/assets/9104402/16768952/ca15c4c6-4846-11e6-8507-818a7102c3a7.png)

Move no. NaN is not really ideal ;-)

dennis-westphal commented:
Offering a draw does not work for both sides:

![screen shot 2016-07-12 at 15 41 18](https://cloud.githubusercontent.com/assets/9104402/16769032/204bd36c-4847-11e6-9765-6533705472b7.png)

dennis-westphal commented:
When a game is won, the claimEther button is displayed. This should not be the cause; the frontend should automatically claim the ether (worked before).

kyroy committed:
Fix bug caused by renaming

kyroy commented:
> the frontend should automatically claim the ether (worked before).

Ye, this worked for onchain... not for offchain ;)

dennis-westphal commented:
Now there seems to be a different problem:

<img width="806" alt="screen shot 2016-07-12 at 19 17 26" src="https://cloud.githubusercontent.com/assets/9104402/16776370/55885a1e-4865-11e6-80db-46ee13d92aeb.png">

dennis-westphal commented:
Also, the Move no NaN issue still exists.

kyroy commented:
Ye, i am working on some problems ;)  
Atm i don't care about the move number... don't know why this was added ;)

kyroy committed:
Fix multiple bugs caused by refactoring or off-chain-chess

Still some bugs existing, just committing my current state

graup commented:
I encountered cases were the move couldn't be saved because the move number wasn't correct. If that's fixed we can remove the counter again :)

kyroy commented:
The move that was send, was send with the old state, not with the new state. This may have caused the error ;)

kyroy committed:
Remove debug output 'move number'

kyroy commented on app/javascript/utils/fen-conversion.js:
```diff
@@ -0,0 +1,251 @@
+  fen += ' 0 ';
+
+  // set fullmove number
+  let halfMoveCounter = state[8] << 7 + state[9];
```
@graup: this new line instead of: `let halfMoveCounter = 128 * state[8] + state[9];` causes problems. Finally found this bug. Normally, we should not try to 'optimize' working functions while refactoring, since new introduced bugs are very hard to trace ;)

graup replied:
Oh no, sorry. That function had another bug (adding En passant flag twice) which I fixed, but apparently then I introduced a new bug.
Maybe translating state to move counter should be its own utility function with a comment, then that mistake couldn't happen.

kyroy committed:
Add update of game variabled, fix generateFen bug

Add update of nextPlayer on move (sent/received)
Move adding of own move to localStorage to games.sendMove
 => since they have to be applied in the correct order
Add update of game.state on eventGameTimeoutStarted if blockchain state newer
Fix bug where generateFen returned wrong move counter -> reason for other bugs

kyroy committed:
Fix wrong variable name in user output

kyroy committed:
Add game.state $watch in playGame.js

Fix enPassant bug in generateState
Remove some console.logs

kyroy committed:
Merge branch 'develop' of github.com:ise-ethereum/on-chain-chess into feature/offchain-frontend-refactor

kyroy committed:
Debug output for board highlighting

Commit:
bugfix move highlighting

Tin-App committed:
Close #210

Added fake move to validate timeout

Tin-App committed:
Merge branch 'feature/offchain-frontend-refactor' of https://github.com/ise-ethereum/on-chain-chess into feature/offchain-frontend-refactor

kyroy committed:
ix bug of wrong moves pushed to gameStates factory

Add update of board, if no move was send to update board function (only state)

kyroy committed:
Fix forgotten BigNumber conversion

Commit:
bugfix mapping problem

Commit:
Merge remote-tracking branch 'origin/feature/offchain-frontend-refactor' into feature/offchain-frontend-refactor

# Conflicts:
#	app/javascript/gamesFactory.js

kyroy committed:
Fix wrong color in ChessLogic.move because of calling by other player
(timeoutState == -2)

kyroy committed:
Merge branch 'feature/offchain-frontend-refactor' of github.com:ise-ethereum/on-chain-chess into feature/offchain-frontend-refactor

graup committed:
Fix showing game state with initial position

graup committed:
Add config.js. Register all accounts with proxy on load, seems to improve reliability

lisasi committed:
PlayGame.html: Add info about Ether

lisasi committed:
Add more info about betting at initGame + fix grammar

lisasi committed:
While waiting for opponent: more info about game

lisasi committed:
Change color of opponents move

Now blue instead of green. Since own move possibilities are also green,
this was hard to distinguish.

lisasi committed:
Change color of selected game in JoinGame for better visability

graup committed:
Fix tests

graup commented:
Proxy communication seems to work (better than before).

The last big bug is that `moveFromState` goes out-of-gas. Still haven't found a good way to debug that :(

lisasi commented on static/playGame.html:
```diff
@@ -15,22 +15,22 @@
-      <span data-ng-show="getGame().pot > 0">The pot of {{getGame().pot}} is being transferred to you.</span>
-      <span data-ng-hide="getGame().pot > 0">The pot was transferred to you.</span>
+      <span data-ng-show="getGame().pot > 0">The pot of {{getGame().pot}} Ether is being transferred to you.</span>
+      <span data-ng-hide="getGame().pot > 0">The pot of {{getGame().pot}} Ether was transferred to you.</span>
```
I added the pot size here, so the users see it after the game has ended. But pot and wonEther are set to zero already, so this does not work. If someone has an idea on how to fix that, that would be nice. Otherwise I change it to the state before.

kyroy replied:
You would have to store it in the contract. Or you need to get it from the old blocks/transactions.

dennis-westphal replied:
This is why it was not displayed at first...

lisasi commented:
Oh, i always read "another" as "other". Weird. Of course you're right @dennis-westphal!

Tin-App committed:
- added test case for out of gas. But commented it out because of invalid signature throw
- changed the chesslogic to not run out of gas

graup committed:
make test pass

graup committed:
Fix converting FEN to state according to specification

graup committed:
fix bug in contract introduced earlier today

graup committed:
contract cleanup

graup committed:
fix tests. set state according to specs

graup committed:
adding another (passing) test case

lisasi committed:
Rechtsschreibfehler korrigiert

dennis-westphal commented on app/less/table.less:
```diff
@@ -9,6 +9,6 @@ tbody {
   }
   tr.selected {
-    background-color:rgba(0, 30, 0, 0.8);
+    background-color:rgba(129, 212, 250, 0.8);
```
Does not improve visability but makes it worse - espacially since the link now is almost invisible. Has to be removed again.

lisasi committed:
Change color of selected game + Fix displayed wonEther in case of draw

graup commented:
Merging to clean up Issues. Going to create a new issue for the remaining bug(s).
